torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -372,20 +371,9 @@ def speculate_subgraph(
                 else contextlib.nullcontext()
             )
 
-            if restore_side_effects:
-                prev_side_effects = tx.output.side_effects.clone()
-
             with autograd_ctx:
                 output = f.call_function(tx, args, sub_kwargs)
 
-            if restore_side_effects:
-                # Captured variables are tracked in side-effects
-                # and they show up in output graph incorrectly.
-                # It is ok to undo this side-effect tracking
-                # as speculate_subgraph will allow only
-                # pure functions.
-                tx.output.side_effects = prev_side_effects
```

Comment:
I don't think this is a complete solution. When we trace through a function capturing a variable, it drops something into SideEffects. That new SideEffect needs to be removed.

Reply:
concretely, it looks like when this happens, we construt a cellvar and then write to it? https://github.com/pytorch/pytorch/blob/d41cfc92e6a19fd3e2a1997882c8535b2a1c3717/torch/_dynamo/variables/functions.py#L45-L49

Reply:
Yes, it does drop something into SideEffects. But if it is read-only, can it cause issues?

If its about hygiene, maybe we can restore just once after both true and false branches have been subtraced?

Reply:
The cellvar isn't read-only though, we construct the cellvar, and then we write to it (via store_cell). Does Dynamo end up reconstructing the cellvar on a graph break?

Reply:
This could possibly lead to issues if we have a subsequent nested scope referring to the same var even though that var hasn't been declared in that scope.

Reply:
I looked little bit more into this. 

1) This code only touches code.co_cellvars which are local variables that are used for the nested function. In cond (or other HOOs), this function will be `true_fn` or `false_fn`. As a result, the store_attr will be called only for the locals (and not nonlocals or globals) here.
https://github.com/pytorch/pytorch/blob/d41cfc92e6a19fd3e2a1997882c8535b2a1c3717/torch/_dynamo/variables/functions.py#L45-L50


2) It is still true that the `side_effects` data structure still has `NewCellVariable` even though they are local to the `true_fn` or `false_fn` (if we don't restore). This could be dangerous because Dynamo can generate bytecode for handling mutation of these variables at the end of tracing parent function. But I checked that Dynamo is smart here, before applying any mutations, it prunes away dead locals here - https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/side_effects.py#L290-L322. As a result, all the mutations that are completely local to subtraced functions are pruned away.


I believe we are still safe here. For any mutation that affects nonlocals or globals while subtracing, our infra will mark it as unsafe here and graph break - https://github.com/pytorch/pytorch/blob/d41cfc92e6a19fd3e2a1997882c8535b2a1c3717/torch/_dynamo/variables/base.py#L106-L115